Implemented first (proper) version or roleplay economy plugin.

GTG3000 opened the pull request:
Features:
- Keeps track of character sheets, including:
  - Fully customiseable fields, meant to store misc data
  - Character inventories (including a "key" inventory that characters can not edit on their own)
  - Money
  - A single image per character
  - Owner of the character may transfer and sell items at whim.
  - Integration with roleplay plugin allows for easy creation of characters to be filled out.
- An item shop, enabling characters to purchase/sell items.
  - Support for fully customizeable unique items to be given to characters, without cluttering up the shop.

Commands:

User commands:
- Char
- Shop
- ShopCategories
- ShopItems
- ItemInfo
- DumpItem

Moderator commands:
- ListChars
- ListItems
- Give
- Take
- GiveCustom
- UploadChar
- UploadItem
- DeleteChar
- DeleteItem

Maintainer commands:
- ReloadEconChars
- ReloadEconItems

GTG3000 committed:
Implemented first (proper) version or roleplay economy plugin.

Features:
- Keeps track of character sheets, including:
  - Fully customiseable fields, meant to store misc data
  - Character inventories (including a "key" inventory that characters can not edit on their own)
  - Money
  - A single image per character
  - Owner of the character may transfer and sell items at whim.
  - Integration with roleplay plugin allows for easy creation of characters to be filled out.
- An item shop, enabling characters to purchase/sell items.
  - Support for fully customizeable unique items to be given to characters, without cluttering up the shop.

Commands:

User commands:
- Char
- Shop
- ShopCategories
- ShopItems
- ItemInfo
- DumpItem

Moderator commands:
- ListChars
- ListItems
- Give
- Take
- GiveCustom
- UploadChar
- UploadItem
- DeleteChar
- DeleteItem

Maintainer commands:
- ReloadEconChars
- ReloadEconItems

rubellyte commented on roleplay_economy.py:
```diff
@@ -0,0 +1,1098 @@
+class RoleplayEconomy(BasePlugin):
+    name = "roleplay_economy"
+    version = "1.0"
+    author = "GTG3000"
```
Don't forget your `description` metafield

rubellyte commented on roleplay_economy.py:
```diff
@@ -0,0 +1,1098 @@
+        "image": None,
+        "inshop": False,
+        "p_buy": 100,
+        "p_sell": 100,
```
I feel like these should probably be a bit more verbose, something like "buy_price" and "sell_price"

rubellyte commented on roleplay_economy.py:
```diff
@@ -0,0 +1,1098 @@
+        _chars = None
+        shop = None
+        gid = None
+        id = None
```
I feel like these should be instance attributes rather than class attributes

rubellyte commented on roleplay_economy.py:
```diff
@@ -0,0 +1,1098 @@
+                               if v]}
+
+        def embed(self):
+            t_embed = Embed(type="rich", colour=16711680)
```
Perhaps make the color int here a hex literal for easier interpretation?

rubellyte commented on roleplay_economy.py:
```diff
@@ -0,0 +1,1098 @@
+            for (f_name, f_data) in self.fields:
+                t_embed.add_field(name=f_name, value=f_data, inline=False)
+
+            t_str = ""
```
More verbose variable name please

rubellyte replied:
Some comments in this upcoming block to explain what you're doing would also be helpful, it's not particularly readable

rubellyte commented on roleplay_economy.py:
```diff
@@ -0,0 +1,1098 @@
+        _items: list
+
+        _emoji = "◀🇽▶"
+        _len = 10
```
Think these should be instance attributes, same as above

rubellyte replied:
although perhaps not _emoji and _len

GTG3000 replied:
For all intents and purposes, those are type hints.

rubellyte commented on roleplay_economy.py:
```diff
@@ -0,0 +1,1098 @@
+        :return:
+        """
+
+        item_embed = Embed(type="rich", colour=16711680)
```
blah blah hex literals

rubellyte commented on roleplay_economy.py:
```diff
@@ -0,0 +1,1098 @@
+
+    def _find_item(self, gid: str, query: str):
+        """
+        Function to finda possible item using sequence matcher.
```
finda

rubellyte commented on roleplay_economy.py:
```diff
@@ -0,0 +1,1098 @@
+            except ValueError:
+                raise CommandSyntaxError(f"{args['pay'][1]} is not a valid amount.")
+
+            recipient = self.Character(self, gid, args['pay'][0].lower())
```
Why is a new Character object being created here?

GTG3000 replied:
Character objects are used as wrappers for interacting with the dictionaries, and do not actually store the data we work with.

I just feel like it improves clarity when you have `recipient.money += x` in stead of `self.chars[gid][args['pay'][0].lower()]['money'] += x`. The upside is that Character object also does verification and initialization for us.

rubellyte replied:
I don't disagree with you on that point, I was just wondering why it was done like so. Seems reasonable.

rubellyte commented on roleplay_economy.py:
```diff
@@ -0,0 +1,1098 @@
+        if args['amount'] < 1:
+            raise CommandSyntaxError("Please use the Take command to remove items.")
+
+        char = self.Character(self, gid, args['name'])
```
Same concern as above

rubellyte commented on roleplay_economy.py:
```diff
@@ -0,0 +1,1098 @@
+        if args['amount'] < 1:
+            raise CommandSyntaxError("Please use the Give command to add items.")
+
+        char = self.Character(self, gid, args['name'])
```
^

rubellyte commented on roleplay_economy.py:
```diff
@@ -0,0 +1,1098 @@
+            except Exception as e:
+                raise CommandSyntaxError(f"Not a valid JSON block: {e}.\n")
+
+        char = self.Character(self, gid, char.lower())
```
^

GTG3000 committed:
Implemented requested changes.
+ Bug fix: shop now properly calculates page count.
+ shop now closes if you try to open another one.
